### PR TITLE
Fix memory leak problem when training on MPS device

### LINKED ISSUE #22621: Potential memory leak when I train the model on my Mac
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

When I tried training a YOLOv11n model on the Pascal VOC 2012 dataset, I noticed that my Mac's memory usage increased linearly with the number of training epochs, even though the terminal logs didn't reflect any memory requests during model training. I suspect this is due to a memory leak, causing the model training to slow down in later stages because of the use of virtual memory.
`

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       2/25      5.44G     0.8462      1.056      1.185         16        640: 100% ━━━━━━━━━━━━ 358/358 1.5it/s 4:04
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 7/7 0.3it/s 20.2s
                   all        200        495      0.808      0.646      0.764        0.6

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       3/25      5.43G      0.944      1.231      1.256         27        640: 100% ━━━━━━━━━━━━ 358/358 1.3it/s 4:37
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 7/7 0.9it/s 8.0s
                   all        200        495      0.691      0.515      0.627      0.508

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       4/25      5.44G     0.9495      1.257      1.257         21        640: 100% ━━━━━━━━━━━━ 358/358 1.2it/s 5:01
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 7/7 1.1it/s 6.5s
                   all        200        495      0.835      0.513      0.698      0.539

`
Here's my memory indicator graph. You can see that the percentage of virtual memory (yellow) is continuously increasing over time.It shows PyCharm requests about 30G RAM during the training, but model only request about 5G RAM in CLI. 

Image

### Environment

Ultralytics 8.3.225 🚀 Python-3.13.9 torch-2.9.0 CPU (Apple M3 Max)
Setup complete ✅ (14 CPUs, 36.0 GB RAM, 442.2/926.4 GB disk)
OS                     macOS-26.1-arm64-arm-64bit-Mach-O
Environment            Darwin
Python                 3.13.9
Install                pip
Path                   /Users/aurxs/.conda/envs/learning/lib/python3.13/site-packages/ultralytics
RAM                    36.00 GB
Disk                   442.2/926.4 GB
CPU                    Apple M3 Max
CPU count              14
GPU                    None
GPU count              None
CUDA                   None
numpy                  ✅ 2.2.6>=1.23.0
matplotlib             ✅ 3.10.7>=3.3.0
opencv-python          ✅ 4.12.0.88>=4.6.0
pillow                 ✅ 12.0.0>=7.1.2
pyyaml                 ✅ 6.0.3>=5.3.1
requests               ✅ 2.32.5>=2.23.0
scipy                  ✅ 1.16.3>=1.4.1
torch                  ✅ 2.9.0>=1.8.0
torch                  ✅ 2.9.0!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.24.0>=0.9.0
psutil                 ✅ 7.1.3
polars                 ✅ 1.35.1
ultralytics-thop       ✅ 2.0.18>=2.0.18
{'OS': 'macOS-26.1-arm64-arm-64bit-Mach-O', 'Environment': 'Darwin', 'Python': '3.13.9', 'Install': 'pip', 'Path': '/Users/aurxs/.conda/envs/learning/lib/python3.13/site-packages/ultralytics', 'RAM': '36.00 GB', 'Disk': '442.2/926.4 GB', 'CPU': 'Apple M3 Max', 'CPU count': 14, 'GPU': None, 'GPU count': None, 'CUDA': None, 'Package Info': {'numpy': '✅ 2.2.6>=1.23.0', 'matplotlib': '✅ 3.10.7>=3.3.0', 'opencv-python': '✅ 4.12.0.88>=4.6.0', 'pillow': '✅ 12.0.0>=7.1.2', 'pyyaml': '✅ 6.0.3>=5.3.1', 'requests': '✅ 2.32.5>=2.23.0', 'scipy': '✅ 1.16.3>=1.4.1', 'torch': '✅ 2.9.0!=2.4.0,>=1.8.0; sys_platform == "win32"', 'torchvision': '✅ 0.24.0>=0.9.0', 'psutil': '✅ 7.1.3', 'polars': '✅ 1.35.1', 'ultralytics-thop': '✅ 2.0.18>=2.0.18'}}

### Minimal Reproducible Example

`
from selenium.webdriver.common.devtools.v140.debugger import resume
from ultralytics import YOLO
import torch
import os

if __name__ == '__main__':

    # --- 1. 确认设备 ---
    if torch.cuda.is_available():
        device = 'cuda'
    elif torch.backends.mps.is_available():
        device = 'mps'
    else:
        device = 'cpu'
    print(f"训练设备自动选择成功: {device}")

    # --- 2. 定义训练参数 ---
    PROJECT_NAME = 'training_runs'
    EXPERIMENT_NAME = 'VOC'

    # --- 3. 加载预训练模型 ---
    print("正在加载预训练模型 'yolov11n.pt'...")
    if os.path.exists(f"./training_runs/{EXPERIMENT_NAME}/weights/last.pt"):
        PATHLAST = f"./training_runs/{EXPERIMENT_NAME}/weights/last.pt"
        RESUME = True
        print("检测到已有训练权重，继续从上次训练点开始训练。")
    else:
        PATHLAST = 'yolo11n.pt'
        RESUME = False
        print("未检测到已有训练权重，使用预训练模型开始训练。")
    model = YOLO(PATHLAST)

    # --- 4. 配置并开始训练 ---
    print("准备开始模型训练...")
    model.train(
        data="VOC.yaml",
        epochs=25,
        batch=-1,
        imgsz=640,
        project=PROJECT_NAME,
        name=EXPERIMENT_NAME,
        exist_ok=True,
        device=device,
        save_period=5,
        conf=0.5,
        workers=0,
        resume=RESUME,
        cache='disk'
    )

    print("\n训练成功完成！")

    # --- 5. 获取最佳模型路径 ---
    save_directory = os.path.join(PROJECT_NAME, EXPERIMENT_NAME)
    best_model_path = os.path.join(save_directory, 'weights/best.pt')

    print(f"最佳模型已保存至: {best_model_path}")

    # ======================================================================
    # ★★★ 核心修改：从预测单张图片升级为遍历整个文件夹 ★★★
    # ======================================================================
    print("\n开始对整个测试集进行批量预测...")

    # --- 6.1 定义测试图片所在的文件夹 ---
    test_images_dir = 'datasets/yomiya/train/images/'

    # --- 6.2 检查模型和文件夹是否存在 ---
    if not os.path.exists(best_model_path):
        print(f"错误：找不到训练好的模型文件 '{best_model_path}'")
    elif not os.path.isdir(test_images_dir):
        print(f"错误：找不到测试图片文件夹 '{test_images_dir}'")
    else:
        # --- 6.3 加载我们训练好的模型 ---
        print("正在加载训练好的最佳模型...")
        inference_model = YOLO(best_model_path)

        # --- 6.4 遍历文件夹中的所有图片并进行预测 ---
        # os.listdir() 获取文件夹下所有文件名
        # 我们只处理以常见图片格式结尾的文件
        image_files = [f for f in os.listdir(test_images_dir) if f.lower().endswith(('.png', '.jpg', '.jpeg'))]

        if not image_files:
            print(f"文件夹 '{test_images_dir}' 中没有找到任何图片文件。")
        else:
            print(f"在 '{test_images_dir}' 中找到 {len(image_files)} 张图片，开始逐一预测...")

            # 我们可以直接把整个文件夹路径传给predict函数，它会自动处理！
            # 这是YOLOv8最方便的功能之一。
            inference_model.predict(
                source=test_images_dir,  # <-- 直接提供文件夹路径
                save=True,  # 保存带有边界框的结果图片
                project=PROJECT_NAME,  # 结果保存在同一个项目下
                name=EXPERIMENT_NAME,  # 结果保存在同一个实验目录下
                exist_ok=True  # 如果目录已存在，则覆盖
            )

            print("\n批量预测完成！")
            print(f"所有预测结果图片已保存在 '{save_directory}' 目录下。")
`

### Additional

_No response_

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by Aurxs**
Fixes #22621
I fix memory leak problem when training on MPS device.

First I add memory reclamation strategy after optimizer.step() and each batch.This may mitigate memory leaks by empty torch cache and recycle useless memory.
Second I noticed that the `self.loss_items` tensor can be separated from the current computation graph which can reduced memory usage without any effect on backpropagation.

I also tested the repair results through a test:
before the fix：
[before_fix.txt](https://github.com/user-attachments/files/23562442/before_fix.txt)
after the fix:
[after_fix.txt](https://github.com/user-attachments/files/23562444/after_fix.txt)
You can see that the maximum memory usage has been significantly reduced after the fix.

I hope my fixes will make some contribution to the community.

I have read the CLA Document and I sign the CLA

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Aurxs, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated review assistant, and a human engineer will be with you shortly. In the meantime, please review the checklist below to help ensure a smooth integration:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/engine/trainer.py [hunk 1]
```diff
@@ -428,6 +428,7 @@ def _do_train(self):
                     self.loss = loss.sum()
                     if RANK != -1:
                         self.loss *= self.world_size
+                    self.loss_items = self.loss_items.detach() # detach to prevent memory leaks
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py]: ❗ **CRITICAL**: Calling `detach()` unconditionally on `self.loss_items` assumes it is always a `torch.Tensor`. In several trainers (e.g., detection) loss items are converted to Python floats or lists for logging, which would cause an `AttributeError` here (`list`/`float` has no `detach`). This change may break non-MPS or non-tensor use cases. Consider guarding the detach by type so it only runs when `self.loss_items` is actually a tensor.

**Suggested change:**
```suggestion
                    self.loss_items = self.loss_items.detach() if isinstance(self.loss_items, torch.Tensor) else self.loss_items  # detach to prevent memory leaks when tensor
```

File: ultralytics/engine/trainer.py [hunk 2]
```diff
@@ -436,6 +437,15 @@ def _do_train(self):
...
                     last_opt_step = ni
 
+                    # MPS-specific memory cleanup after optimizer step
+                    if self.device.type == "mps":
+                        torch.mps.synchronize()  # Ensure all operations complete before cleanup
+                        torch.mps.empty_cache()
+                        # More aggressive GC in early epochs to prevent initial memory buildup
+                        gc_interval = 1 if epoch < 5 else (5 if epoch < 10 else 10)
+                        if (ni + 1) % gc_interval == 0:
+                            gc.collect()
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py:447]: ⚠️ **HIGH**: `gc.collect()` is used here but `gc` is not imported anywhere in this file in the current repo version. This will raise a `NameError` the first time this code path is hit, including on CUDA/CPU if `device.type` is ever "mps". Please ensure `import gc` is added at the top of this module or reuse an existing import.

File: ultralytics/engine/trainer.py [hunk 3]
```diff
@@ -465,6 +475,15 @@ def _do_train(self):
...
                 self.run_callbacks("on_train_batch_end")
 
+                # MPS-specific: Clear batch data to prevent accumulation
+                # More aggressive cleanup in early epochs
+                if self.device.type == "mps":
+                    batch_interval = 1 if nb < 5 else (5 if nb < 25 else 25)
+                    if (i + 1) % batch_interval == 0:
+                        gc.collect()
+                        torch.mps.synchronize()  # Ensure all operations complete before cleanup
+                        torch.mps.empty_cache()
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py:485]: ⚠️ **HIGH**: Same as the optimizer-step block above, this batch-level cleanup uses `gc.collect()` without guaranteeing that `gc` is imported, which will cause a `NameError` when this path is executed on MPS. Once `import gc` is added, this will be resolved.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Main concerns: (1) `self.loss_items.detach()` is applied unconditionally and may break trainers where `loss_items` is a list/float rather than a tensor; guard by type. (2) `gc.collect()` is introduced in two places but `gc` is not imported, which will cause a `NameError` on MPS. (3) The new MPS cleanup blocks are quite aggressive and hard-coded; consider making their intervals configurable or less frequent to avoid unnecessary performance degradation while still addressing memory issues.

💬 Posted 3 inline comments
